extends Node2D

@onready var slot_mensagem_scene := preload("res://cenas/mensagem_slot.tscn")
@onready var slot_mensagem_scene_2 := preload("res://cenas/online_users.tscn")
@onready var container_mensagens := $chat/ChatGlobal1/ScrollContainer_chat_global/VBoxContainer
@onready var container_mensagens_2 := $chat/online_layer/ScrollContainer_usuarios_online/VBoxContainer
@onready var anim_player := $chat/AnimationPlayer
@onready var timer := $Timer
@onready var arrastavel_layer := $layer_arratavel_chat
@onready var arrastavel_online := $layer_arratavel_online

var online := false
var arrastando := false
var offset := Vector2.ZERO
var pode_arrastar := false

var chat_maximizado := true
var online_visivel := true
var cascata_minimizar_online = false
var qual_ta_rodando = ""
# ----------------- READY ------------------

func _ready():
	
	$chat/ChatGlobal1/chat_global2/mandar_mensagem.text = ""
	
	if online:
		Socket.connect("server_receive", get_message)
		Socket.connect("server_receive", get_user_connected)

		for player in Global.players.values():
			print('Adicionando nick ', player)
			adicionar_mensagem_2(player.username)

		
	else:
		adicionar_mensagem("helcio", "Mensagem offline de teste (recebida)", false)
		adicionar_mensagem("Voc√™", "Mensagem offline de teste (enviada)", true)
		adicionar_mensagem_2("helcio - nick ")
		adicionar_mensagem_2("Voc√™")

	$chat/ChatGlobal1/chat_global2/mandar_mensagem.connect("text_submitted", Callable(self, "_enviar_mensagem"))
	anim_player.animation_finished.connect(_on_animation_finished)


# 	--------------- MENSAGENS ------------------

func get_message(flag, response):
	if flag != "get_message":
		return
	var nome = response.get("username", "Desconhecido")
	var texto = response.get("message", "")
	adicionar_mensagem(nome, texto, false)
	print("üí¨ Mensagem recebida:", response)
	
func get_user_connected(flag, response):
	if flag != "_new_player_connected":
		return


	if Global.players.has(response.keys()[0]) or response.keys()[0] == Sessao.id:
		return

	var user = response.keys()[0]

	Global.players[user] = response[user]
	adicionar_mensagem_2(response[user].username)

	adicionar_mensagem(response.username, response.message, false)

	print("üí¨ Mensagem recebida:", response)


func _enviar_mensagem():
	var chat_enviar = $chat/ChatGlobal1/chat_global2/mandar_mensagem/texto_mensagem.text.strip_edges()
	if chat_enviar == "":
		return

	var nome = Sessao.nick if Sessao.has_meta("nick") else "Voc√™"

	if online:
		Socket.enviar_json("enviar_mensagem_chat", {
			"id": Sessao.id,
			"chat_enviar": chat_enviar
		})

	# Adiciona localmente sempre, mesmo offline
	adicionar_mensagem(nome, chat_enviar, true)

	$chat/ChatGlobal1/chat_global2/mandar_mensagem/texto_mensagem.text = ""



func adicionar_mensagem(nick, texto, enviada_por_mim := false):
	var novo_slot = slot_mensagem_scene.instantiate()
	novo_slot.get_node("VBoxContainer/HBoxContainer/nick").text = nick
	novo_slot.get_node("VBoxContainer/texto").text = texto
	novo_slot.configurar_espaco(enviada_por_mim)
	container_mensagens.add_child(novo_slot)
	
func adicionar_mensagem_2(nick):
	var novo_slot = slot_mensagem_scene_2.instantiate()
	novo_slot.start(nick, container_mensagens_2)


# ----------------- ENVIO ------------------

func _on_enviar_botao_pressed():
	_enviar_mensagem()

func _on_mandar_mensagem_pressed():
	_enviar_mensagem()


# ----------------- ARRASTO ------------------

func _input(event: InputEvent) -> void:
	if event is InputEventMouseButton and event.button_index == MOUSE_BUTTON_LEFT:
		if event.pressed and pode_arrastar:
			arrastando = true
			offset = global_position - get_global_mouse_position()
		else:
			arrastando = false
	if event is InputEventKey and event.pressed and event.keycode == KEY_ENTER:
		_enviar_mensagem()
	elif event is InputEventMouseMotion and arrastando:
		global_position = get_global_mouse_position() + offset

func _on_painel_arrasto_mouse_entered():
	Global.mouse_sobre_chat = true
	pode_arrastar = true

func _on_painel_arrasto_mouse_exited():
	Global.mouse_sobre_chat = false
	pode_arrastar = false


# ----------------- ANIMA√á√ïES CHAT ------------------

func _on_minimizar_pressed() -> void:
	if anim_player.is_playing() or not chat_maximizado and not cascata_minimizar_online:
		return
	if chat_maximizado and not cascata_minimizar_online:
		anim_player.play("usuarios_online_fechar")
		qual_ta_rodando = "abrir"
		$Timer.start()
		print("ü°á Minimizar Chat")
		return

	if chat_maximizado and cascata_minimizar_online:
		anim_player.play("usuarios_online_fechar_completo")
		qual_ta_rodando = "abrir"
		$Timer.start()
		print("ü°á Minimizar Chat")
		return


func _on_maximizar_2_pressed() -> void:
	if anim_player.is_playing() or chat_maximizado and not cascata_minimizar_online:
		return
	if not chat_maximizado and not cascata_minimizar_online:
		#$chat/online_layer.visible = true
		anim_player.play("fechar")
		qual_ta_rodando = "usuarios_online_fechar_completo_max"
		$Timer.start()
		print("ü°á Minimizar Chat")
		return
	if not chat_maximizado and cascata_minimizar_online:
		anim_player.play("fechar")
		qual_ta_rodando = "usuarios_online_fechar_completo_max_diferente"
		$Timer.start()
		print("ü°á Minimizar Chat")
		return
	

func _on_minimizar_user_online_pressed() -> void:
	if anim_player.is_playing() or not online_visivel:
		return

	anim_player.play("usuarios_online_fechar")
	print("ü°á Minimizar Online")
	cascata_minimizar_online = true


func _on_maximizar_user_online_pressed() -> void:
	if anim_player.is_playing() or online_visivel:
		return

	anim_player.play("usuarios_online_abrir")
	print("ü°Ö Maximizar Online")
	cascata_minimizar_online = false

# ----------------- FIM DAS ANIMA√á√ïES ------------------

func _on_animation_finished(anim_name: String) -> void:
	print("üé¨ Finalizou anima√ß√£o:", anim_name)

	match anim_name:
		"abrir":
			chat_maximizado = false
			arrastavel_layer.visible = false
			arrastavel_online.visible = false
			print("üîí Chat minimizado")

		"fechar":
			chat_maximizado = true
			arrastavel_layer.visible = true
			if online_visivel:
				arrastavel_online.visible = true
			print("üîì Chat maximizado")

		"usuarios_online_fechar":
			online_visivel = false
			arrastavel_online.visible = false
			print("üë• Online ocultado")

		"usuarios_online_abrir":
			online_visivel = true
			arrastavel_online.visible = true
			print("üë• Online vis√≠vel")
			
		"usuarios_online_fechar_completo":
			online_visivel = true
			arrastavel_online.visible = true
			print("üë• Online vis√≠vel")

func _on_timer_timeout() -> void:
	if qual_ta_rodando == "":
		return

	if not anim_player.is_playing():
		if not cascata_minimizar_online and qual_ta_rodando == "abrir":
			anim_player.play("usuarios_online_fechar_completo")
			$time2.start()
	if not anim_player.is_playing():
		if not anim_player.is_playing() and qual_ta_rodando == "usuarios_online_fechar_completo_max":
			anim_player.play_backwards("usuarios_online_fechar_completo")
			$time2.start()
	if not anim_player.is_playing():
		if not anim_player.is_playing() and qual_ta_rodando == "usuarios_online_fechar_completo_max_diferente":
			anim_player.play_backwards("usuarios_online_fechar_completo")
			$time2.start()
	if not anim_player.is_playing():
		if cascata_minimizar_online:
			anim_player.play("abrir")
			$chat/online_layer.visible = false
			qual_ta_rodando = "" # Limpa ap√≥s rodar
	#if not anim_player.is_playing():
		#if cascata_minimizar_online:
			##anim_player.play("fechar")
			#$chat/online_layer.visible = true
			#qual_ta_rodando = ""  # Limpa ap√≥s rodar
			
			
func _on_time_2_timeout() -> void:
	if qual_ta_rodando != "" and qual_ta_rodando == "abrir":
		$chat/online_layer.visible = false
		anim_player.play(qual_ta_rodando)
		print("‚ñ∂ Rodando via time_2:", qual_ta_rodando)
		qual_ta_rodando = "" # Limpa ap√≥s rodar
	if qual_ta_rodando != "" and qual_ta_rodando == "usuarios_online_fechar_completo_max":
		$chat/online_layer.visible = true
		anim_player.play("usuarios_online_abrir")
		print("‚ñ∂ Rodando via time_2:", qual_ta_rodando)
		qual_ta_rodando = "" # Limpa ap√≥s rodar
	if qual_ta_rodando != "" and qual_ta_rodando == "usuarios_online_fechar_completo_max_diferente":
		$chat/online_layer.visible = true
		#anim_player.play("usuarios_online_abrir")
		print("‚ñ∂ Rodando via time_2:", qual_ta_rodando)
		qual_ta_rodando = ""
	#if qual_ta_rodando != "" and qual_ta_rodando == "usuarios_online_fechar_completo_max":
		#$chat/online_layer.visible = true
		#anim_player.play("usuarios_online_abrir")
		#print("‚ñ∂ Rodando via time_2:", qual_ta_rodando)
		#qual_ta_rodando = ""  # Limpa ap√≥s rodar

	$time2.stop() # Evita loop infinito
